src/main/java/net/minecraftforge/common/data/JsonCodecProvider.java

```
@@ -0,0 +1,103 @@
+     * @param registryKey ResourceKey identifying the registry and its directory.
+     * @param entries Map of entries to encode and their ResourceLocations. Paths for values are derived from the ResourceLocation's entryid:entrypath.
+     */
+    public static <T> JsonCodecProvider<T> forDatapackRegistry(DataGenerator dataGenerator, ExistingFileHelper existingFileHelper, String modid, RegistryOps<JsonElement> registryOps, ResourceKey<Registry<T>> registryKey, Map<ResourceLocation, T> entries)
+    {
+        final ResourceLocation registryId = registryKey.location();
+        // Minecraft datapack registry folders are in data/json-namespace/registry-name/
+        // Non-vanilla registry folders are data/json-namespace/registry-namespace/registry-name/
+        final String registryFolder = registryId.getNamespace().equals("minecraft")
+            ? registryId.getPath()
+            : registryId.getNamespace() + "/" + registryId.getPath();
+        final Codec<T> codec = (Codec<T>)RegistryAccess.REGISTRIES.get(registryKey).codec();
+        return new JsonCodecProvider<>(dataGenerator, existingFileHelper, modid, registryOps, PackType.SERVER_DATA, registryFolder, codec, entries);
+    }
+
+    @Override
+    public void run(final CachedOutput cache) throws IOException
+    {
+        final Path outputFolder = this.dataGenerator().getOutputFolder();
+        final String dataFolder = this.packType().getDirectory();
+        for (final var entry : this.entries().entrySet())
+        {
+            final ResourceLocation id = entry.getKey();
+            final T value = entry.getValue();
+            final Path path = outputFolder.resolve(String.join("/", dataFolder, id.getNamespace(), this.directory(), id.getPath()+".json"));
+            this.codec().encodeStart(this.dynamicOps(), value)
+                .resultOrPartial(msg -> LOGGER.error("Failed to encode {}: {}", path, msg)) // Log error on encode failure.
```

Should we consider throwing an exception instead of logging the error?

---

yes